File: urbanairship-layout/src/main/java/com/urbanairship/android/layout/property/ToggleStyle.java
/* Copyright Airship and Contributors */

package com.urbanairship.android.layout.property;

import com.urbanairship.json.JsonException;
import com.urbanairship.json.JsonMap;

import androidx.annotation.NonNull;

public abstract class ToggleStyle {
    @NonNull
    private final ToggleType type;


    ToggleStyle(@NonNull ToggleType type) {
        this.type = type;
    }

    @NonNull
    public static ToggleStyle fromJson(@NonNull JsonMap json) throws JsonException {
        String typeString = json.opt("type").optString();
        switch (ToggleType.from(typeString)) {
            case SWITCH:
                return SwitchStyle.fromJson(json);
            case CHECKBOX:
                return CheckboxStyle.fromJson(json);
        }
        throw new JsonException("Failed to parse ToggleStyle! Unknown type: " + typeString);
    }

    @NonNull
    public ToggleType getType() {
        return type;
    }
}

File: urbanairship-layout/src/main/java/com/urbanairship/android/layout/property/ToggleStyle.kt
/* Copyright Airship and Contributors */
package com.urbanairship.android.layout.property

import com.urbanairship.json.JsonException
import com.urbanairship.json.JsonValue

public abstract class ToggleStyle internal constructor(
    public val type: ToggleType
) {

    public companion object {
        private const val KEY_TYPE = "type"

        @Throws(JsonException::class)
        public fun fromJson(json: JsonValue): ToggleStyle {
            val content = json.requireMap()
            val type = ToggleType.from(content.require(KEY_TYPE))
            return when (type) {
                ToggleType.SWITCH -> SwitchStyle.fromJson(json)
                ToggleType.CHECKBOX -> CheckboxStyle.fromJson(json)
            }
        }
    }
}
